nova/scheduler/filters/resize_reservation_filter.py

```diff
@@ -0,0 +1,72 @@
+# Copyright (c) 2020 SAP SE
+# All Rights Reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+from oslo_log import log as logging
+
+import nova.conf
+from nova.scheduler import filters
+from nova.scheduler import utils
+
+LOG = logging.getLogger(__name__)
+
+CONF = nova.conf.CONF
+
+
+class ResizeReservedRAMFilter(filters.BaseHostFilter):
+    """Only return hosts for resize with sufficient available RAM."""
+
+    RUN_ON_REBUILD = False
+
+    def reserved_percent_ram(self):
+        return CONF.filter_scheduler.resized_reserved_ram_percent
+
+    def host_passes(self, host_state, request_spec):
+        if utils.request_is_resize(request_spec):
+            return True
+
+        reserved_ram_ratio = self.reserved_percent_ram() / 100.0
+        reserved_ram = host_state.total_usable_ram_mb * reserved_ram_ratio
+
+        # Apply reserved memory to total and free memory
+        host_state.total_usable_ram_mb -= reserved_ram
+        free_ram_mb = host_state.free_ram_mb - reserved_ram
+
+        memory_mb_limit = \
+            host_state.total_usable_ram_mb * host_state.ram_allocation_ratio
+        used_ram_mb = host_state.total_usable_ram_mb - free_ram_mb
+        usable_ram = memory_mb_limit - used_ram_mb
```

Comment:
As I understand it: `total_usable_ram_mb` doesn't include the `ram_allocation_ratio`. Does `free_ram_mb` include the allocation ratioed RAM? Is it always negative if the `ram_allocation_ratio` is > 1.0 and the host is full?

Could we simplify this whole operation by saying that we reserve X % of the overcommited RAM?
```python
memory_mb_limit = total_usable_ram_mb * ram_allocation_ratio
used_ram_mb = total_usable_ram_mb - free_ram_mb
wanted_ram_mb = used_ram_mb + request_spec.memory_mb
wanted_ram_percent = wanted_ram_mb / memory_mb_limit * 100
if 100 - wanted_ram_percent > self.reserved_percent_ram():
    LOG.info('Spawning on host (<host>) would use part of the memory reserved for resizes (X %).
    return False
```

I find this much easier to think about, but might be because I just wrote it. Thoughts?

Reply:
I agree! Will change it!

Reply:
I agree! Will do!